modules/type/hero_power/hero_power.gd

```diff
@@ -0,0 +1,120 @@
+extends Node
+
+
+#region Signals
+## Emits whenever a hero power gets used.
+signal used(after: bool, player: Player, sender_peer_id: int)
+#endregion
+
+
+#region Internal Functions
+func _ready() -> void:
+	Modules.register(&"HeroPower", [&"Type"], func() -> void:
+		# Load module.
+		Modules.register_hooks(&"HeroPower", self.handler)
+		Modules.register_packet(&"HeroPower", &"Hero Power")
+		
+		TypeModule.register_type(&"HeroPower", false)
+	, func() -> void:
+		# Unload module. No need to unregister hooks.
+		TypeModule.unregister_type(&"HeroPower")
+	)
+#endregion
+
+
+#region Public Functions
+func handler(what: Modules.Hook, info: Array) -> bool:
+	if what == Modules.Hook.ANTICHEAT:
+		return anticheat_hook.callv(info)
+	elif what == Modules.Hook.ACCEPT_PACKET:
+		return await accept_packet_hook.callv(info)
+	
+	return true
+
+
+func is_hero_power(card: Card) -> bool:
+	return TypeModule.is_card_type(card, &"HeroPower")
+
+
+func accept_hero_power_packet(player: Player, sender_peer_id: int) -> bool:
+	var hero_power: Card = player.hero.hero_power
+	
+	used.emit(false, player, sender_peer_id)
+	
+	hero_power.refunded = false
+	
+	hero_power.trigger_ability(&"Hero Power", false)
+	if hero_power.refunded:
+		return false
+	
+	player.has_used_hero_power_this_turn = true
+	player.mana -= hero_power.cost
+	
+	used.emit(true, player, sender_peer_id)
+	return true
+
+
+func accept_play_packet(
+	player: Player,
+	sender_peer_id: int,
+	
+	location: StringName,
+	location_index: int,
+	board_index: int,
+	position: Vector3i,
+) -> bool:
+	var card: Card = Card.get_from_index(player, location, location_index)
+	
+	if not is_hero_power(card):
+		return true
+	
+	card.trigger_ability(&"Battlecry", false)
+	await card._wait_for_ability(&"Battlecry")
+	
+	if card.refunded:
+		return false
+	
+	card.location = &"Hero Power"
+	return true
+
+
+#region Hooks
+# Hero Power
+func anticheat_hook(packet_type: StringName, sender_peer_id: int, sender_player: Player, actor_player: Player, info: Array) -> bool:
+	if packet_type != &"Hero Power":
+		return true
+	
+	# The info needs to be correct.
+	if not Anticheat.info_check(info, []):
+		Anticheat.feedback("Invalid hero power info.", sender_peer_id)
+		return false
+	
+	var hero_power: Card = actor_player.hero.hero_power
```

Comment:
I don't think `Player.hero` or `Card.hero_power` should be in core, I'm not sure what the alternative solution is.